ublox_derive/src/output.rs

```diff
@@ -0,0 +1,140 @@
+use crate::types::{PackDesc, PackFieldEnum, PackFieldFlags, PackFieldRepr};
+use proc_macro2::TokenStream;
+use quote::{format_ident, quote};
+use syn::Type;
+
+pub fn generate_recv_code_for_packet(pack_descr: &PackDesc) -> TokenStream {
+    let ref_name = format_ident!("{}Ref", pack_descr.name);
+    let mut getters = Vec::with_capacity(pack_descr.fields.len());
+
+    for f in &pack_descr.fields {
+        let ty = f.intermidiate_type();
+        let get_name = &f.name;
+        getters.push(quote! {
+            #[inline]
+            pub fn #get_name(&self) -> #ty {
+            unimplemented!()
+            }
+        });
+    }
+    quote! {
+    pub struct #ref_name<'a>(&'a [u8]);
+    impl<'a> #ref_name<'a> {
+            #(#getters)*
+    }
+    }
+}
+
+pub fn generate_types_for_packet(pack_descr: &PackDesc) -> TokenStream {
+    let mut req_types = Vec::new();
+    for f in &pack_descr.fields {
+        let ty = f.intermidiate_type();
+        match f.repr {
+            PackFieldRepr::Plain | PackFieldRepr::Map(_) => {}
+            PackFieldRepr::Enum(ref e) => {
+                req_types.push(define_enum(e, ty, &f.ty));
+            }
+            PackFieldRepr::Flags(ref flags) => {
+                req_types.push(define_bitflags(flags, ty, &f.ty));
+            }
+        }
+    }
+    let name = format_ident!("{}Raw", pack_descr.name);
+    let class = pack_descr.header.class;
+    let id = pack_descr.header.id;
+    let fixed_len = match pack_descr.header.fixed_len {
+        Some(x) => quote! { Some(#x) },
+        None => quote! { None },
+    };
+    quote! {
+        #(#req_types)*
+
+    impl UbxPacket for #name {
+        const class: u8 = #class;
+        const id: u8 = #id;
+    const fixed_length: Option<u16> = #fixed_len;
+    }
+    }
+}
+
+pub fn generate_send_code_for_packet(pack_descr: &PackDesc) -> TokenStream {
+    let payload_struct = format_ident!("{}Payload", pack_descr.name);
+
+    let mut fields = Vec::with_capacity(pack_descr.fields.len());
+    for f in &pack_descr.fields {
+        let ty = f.intermidiate_type();
+        let name = &f.name;
+        fields.push(quote! {
+            #name: #ty
+        });
+    }
+
+    match pack_descr.packet_size() {
+        Some(packet_size) => {
+            quote! {
+            pub struct #payload_struct {
+                #(#fields),*
+            }
+            impl #payload_struct {
+                #[inline]
+                fn to_packet(self) -> [u8; #packet_size] {
+                unimplemented!();
+                }
+            }
+            }
+        }
+        None => {
+            quote! {
+            pub struct #payload_struct {
+                #(#fields)*,
+            }
+            impl #payload_struct {
+                #[inline]
+                fn to_packet(self, out: &mut Vec<u8>) {
```

Comment:
As above, we should be able to take `&mut [u8]` here.

Reply:
> As above, we should be able to take &mut [u8] here.

I should note that this is interface for dynamic-sized packages.
And change to `&mut [u8]` makes it squishy, because of what it  there is no enough space into buffer? In this case we should return error and user should allocate more space,
plus it is Rust, so he/she should zero intialize new allocated bytes or use `unsafe`.

This make interface squishy, while usage of `Vec<u8>` make it super simpliy, we allocate as many space as we need without "the second call" and write data right after allocation without need extra assign to zero.

May be there is no need to send variable-sized packages at all?

Reply:
The user may not be able to allocate more memory, or allocating more memory may be more complicated then expanding a vector. What if the packet had a `get_formatted_size` method that returned the size?

I think some of the AID packets are variable-sized to the chip. Otherwise I can't think of anything off the top of my head.

Reply:
> What if the packet had a get_formatted_size

I suppose  `std::io::Write` will be better, because of `get_formatted_size` can be effective and with bad API:

```rust
impl FixedSizePack {
   const fn get_formatted_size() -> usize;
}

impl VarSizePack1 {
   const fn get_formatted_size(variable_field1_size: usize) -> usize;
}
impl VarSizePack2 {
   const fn get_formatted_size(variable_field2_size: usize, variable_field3_size: usize) -> usize;
}
```

Or not effective with good api:
```rust
trait GetPacketSizeInsideWire {
   fn get_formatted_size(&self) -> usize;
}
```

In the first case user can allocate space on the stack, without VLA (Variable-length array) / Heap,
in the second API is consistent, but user have to use "heavy" methods for memory allocations.

Reply:
I agree, either the user will have to "just allocate enough" on the stack, or will have to use heap options. If we pass a `Vec` then we're forcing the user to use a Vec (I'd have to think more about `std::io::Write`, but as mentioned above, I don't think it's `no_std`).